src/test/java/io/nats/streaming/SubscribeTests.java

```diff
@@ -870,4 +926,47 @@ public void testNoDuplicatesOnSubscriberStart() throws Exception {
             }
         }
     }
+
+    @Test
+    public void testSubscriptionCantBlockHeartbeat() throws Exception {
+        String args[] = {"-hbi", "1s",
+                        "-hbt", "1s",
+                        "-hbf", "1"};
+        try (NatsStreamingTestServer srv = new NatsStreamingTestServer(clusterName, args, false)) {
+            Options options = new Options.Builder().natsUrl(srv.getURI()).build();
+            try (StreamingConnection sc = NatsStreaming.connect(clusterName, clientName, options)) {
+                final int count = 5;
+                final int sleep = 1337; // Nice leet number so we get odd overlaps with HB
+
+                for (int i = 1; i <= count; i++) {
+                    byte[] data = String.format("%d", i).getBytes();
+                    sc.publish("foo", data);
+                }
+                sc.getNatsConnection().flush(Duration.ofSeconds(1));
+
+                final CountDownLatch latch = new CountDownLatch(1);
+                final AtomicInteger received = new AtomicInteger(0);
+
+                // Capture the messages that are delivered.
+                MessageHandler mcb = msg -> {
+                    if (received.incrementAndGet() >= count) {
+                        latch.countDown();
+                    }
+                    try {
+                        Thread.sleep(sleep);
+                    } catch (Exception exp) {
+                        //ignore
+                    }
+                };
+
+                // Should receive all messages, and heartbeat should be ok despite sleeps in callback.
+                try (Subscription sub = sc.subscribe("foo", mcb,
+                        new SubscriptionOptions.Builder().deliverAllAvailable().build())) {
+
+                    assertTrue("Did not receive our messages", latch.await(3 * count * sleep, TimeUnit.SECONDS));
+                    assertEquals("Got wrong number of msgs", count, received.get());
+                }
```

Comment:
You are not really checking that the client has not been closed by the server here. So what I would do is initially send a single message, sleep in the callback for more than the server would need to close the client if the HB was blocked by callback.
After that time, send a new message and this time you don't need to sleep in callback. Test is OK if callback received 2 messages.